_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+    updateBackendsAndGet: UpdateBackendsAndGet,
+    healthChecksRoundRobin: HealthChecksRoundRobin,
+    sendAndExpectStatus: SendAndExpect[Status],
+    healthCheckInterval: HealthCheckInterval
+  ): IO[Unit] =
+    (for {
+      currentUrl <- healthChecksRoundRobin(healthChecks)
+      uri        <- IO.fromEither(parseUri(currentUrl.value))
+      status     <- sendAndExpectStatus(uri)
+      _          <- updateBackendsAndGet(backends, currentUrl, status)
+    } yield ())
+      .flatMap(_ => IO.sleep(healthCheckInterval.value.seconds))
+      .foreverM
+}
+```
+
+As you can see `HealthCheckBackends` is just a piece of logic which uses the services which we defined above.
+
+We can move to `LoadBalancer` now.
+
+This will be a service which:
+- accepts any request (let's say `"localhost:8080/users/1"`)
+- applies Round Robin to `Backends` (let's say the current backend will be `"localhost:8083"`)
+- updates uri in the following fashion: `"localhost:8080/users/1"` => `"localhost:8083/users/1"` (so that it hits the correct backend resource)
+- parses uri
+- sends request to uri
+- sends response to the client
+
+```scala
+package com.ghurtchu.loadbalancer.services
+
+import cats.effect.IO
+import com.ghurtchu.loadbalancer.domain.Backends.Backends
+import com.ghurtchu.loadbalancer.services.RoundRobin.BackendsRoundRobin
+import org.http4s.dsl.Http4sDsl
+import org.http4s.{HttpRoutes, Request}
+
+object LoadBalancer {
+
+  def from(
+    backends: Backends,
+    sendAndExpectResponse: Request[IO] => SendAndExpect[String],
+    parseUri: ParseUri,
+    backendsRoundRobin: BackendsRoundRobin,
+  ): HttpRoutes[IO] = {
+    val dsl = new Http4sDsl[IO] {}
+    import dsl._
+    HttpRoutes.of[IO] { case request =>
+      backendsRoundRobin(backends).flatMap {
+        _.fold(Ok("All backends are inactive")) { currentUrl =>
+          val urlUpdated = currentUrl.value
+            .concat(request.uri.path.renderString)
+          for {
+            uri      <- IO.fromEither(parseUri(urlUpdated))
+            response <- sendAndExpectResponse(request)(uri)
+            result   <- Ok(response)
+          } yield result
+        }
+      }
+    }
+  }
+
+  final case class InvalidUri(uri: String) extends Throwable {
+    override def getMessage: String =
+      s"Could not construct proper URI from $uri"
+  }
+}
+```
+Awesome, we defined the `LoadBalancer` which will return `HttpRoutes[IO]` and `http4s` will run it.
+
+Now, let's see how "the end of the world" would look like:
+
+```scala
+import cats.effect.{IO, IOApp}
+import cats.implicits.{catsSyntaxTuple2Parallel, catsSyntaxTuple2Semigroupal}
+import com.comcast.ip4s.{Host, Port}
+import com.ghurtchu.loadbalancer.domain.{Config, Urls}
+import com.ghurtchu.loadbalancer.domain.Backends.{Backends, HealthChecks}
+import com.ghurtchu.loadbalancer.http.HttpServer
+import com.ghurtchu.loadbalancer.services.{ParseUri, RoundRobin, UpdateBackendsAndGet}
+import pureconfig._
+import pureconfig.generic.auto._
```

Comment:
The package is `pureconfig.generic.derivation.default.*`
You can use `derives ConfigReader` in the Config case classes to generate the necessary givens.